### PR TITLE
CodeCamp #140 [Feature] Add synapse dataset and data augmentation in dev-1.x.

### PR DESCRIPTION
**Pull request by Dominic23331**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Support synapse and data augmentation
dataset link: https://www.synapse.org/#!Synapse:syn3193805/wiki/
TransUnet use this dataset, paper link: https://arxiv.org/pdf/2102.04306.pdf

## Modification

Add synapse dataset loader.
Add a python script to transform synapse dataset format to mmseg dataset format.
Add synapse dataset augmentation.

## Update 2022-12-13

The result when using 13 classes:
```python
        classes=('background', 'spleen', 'right_kidney', 'left_kidney',
                 'gallbladder', 'esophagus', 'liver', 'stomach', 'aorta',
                 'inferior_vena_cava', 'portal_vein_and_splenic_vein',
                 'pancreas', 'right_adrenal_gland', 'left_adrenal_gland'),
        palette=[[0, 0, 0], [255, 127, 127], [224, 231, 161], [138, 204, 132],
                 [64, 172, 136], [126, 152, 187], [140, 110, 160],
                 [247, 88, 240], [202, 172, 161], [237, 213, 149],
                 [139, 182, 139], [111, 192, 185], [82, 107, 163],
                 [89, 54, 156]])
```

image

### CODE REVIEW

**Comment by Dominic23331** [comment]
add synapse dataset and it`s augmentations.

**Comment by MengzhangLI** [comment]
Hi, thanks for your nice PR. We would review it asap.

Best wishes,

File: tests/test_datasets/test_dataset.py [hunk 1]
```diff
@@ -220,6 +222,19 @@ def test_vaihingen():
...
+    test_dataset = SynapseDataset(
+        pipeline=[],
+        data_prefix=dict(
+            img_path=osp.join(
+                osp.dirname(__file__),
+                '../data/pseudo_synapse_dataset/img_dir'),
+            seg_map_path=osp.join(
+                osp.dirname(__file__),
+                '../data/pseudo_synapse_dataset/ann_dir')))
+    assert len(test_dataset) == 5
```
Comment (MengzhangLI) [inline tests/test_datasets/test_dataset.py:235]: Not necessary to upload too many files in pesudo dataset folder. I think 2 or 3 is enough.

File: mmseg/datasets/synapse.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+                 'inferior_vena_cava', 'portal_vein_and_splenic_vein',
+                 'pancreas', 'right_adrenal_gland', 'left_adrenal_gland'),
+        palette=[[0, 0, 0], [255, 127, 127], [224, 231, 161], [138, 204, 132],
+                 [64, 172, 136], [126, 152, 187], [140, 110, 160],
+                 [247, 88, 240], [202, 172, 161], [237, 213, 149],
+                 [139, 182, 139], [111, 192, 185], [82, 107, 163],
+                 [89, 54, 156]])
+
+    def __init__(self, **kwargs) -> None:
+        super().__init__(img_suffix='.jpg', seg_map_suffix='.png', **kwargs)
```
Comment (MengzhangLI) [inline mmseg/datasets/synapse.py:26]:
```suggestion
    def __init__(self, 
                img_suffix='.jpg',
                seg_map_suffix='.png',
                **kwargs) -> None:
        super().__init__(img_suffix=img_suffix, seg_map_suffix=seg_map_suffix, **kwargs)
```

Please check out this new PR: https://github.com/open-mmlab/mmsegmentation/pull/2379. Under this modification, the key-word arguments can be allowed as input.

File: tools/dataset_converters/synapse.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+
+    if not osp.exists(osp.join(dataset_path, 'img')) \
+            or not osp.exists(osp.join(dataset_path, 'label')):
+        raise FileNotFoundError('The dataset structure is incorrect.'
+                                ' Please check your dataset.')
+
+    mkdir_or_exist(osp.join(save_path, 'img_dir'))
+    mkdir_or_exist(osp.join(save_path, 'ann_dir'))
+
+    if osp.exists(osp.join(dataset_path, 'train.txt')) \
```
Comment (MengzhangLI) [inline tools/dataset_converters/synapse.py:114]: Could you please provide `train.txt` and `val.txt` which makes the same with TransUNet train/val split? In this default setting, all datasets are handled into training set.

File: mmseg/datasets/synapse.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+                 'pancreas', 'right_adrenal_gland', 'left_adrenal_gland'),
+        palette=[[0, 0, 0], [255, 127, 127], [224, 231, 161], [138, 204, 132],
+                 [64, 172, 136], [126, 152, 187], [140, 110, 160],
+                 [247, 88, 240], [202, 172, 161], [237, 213, 149],
+                 [139, 182, 139], [111, 192, 185], [82, 107, 163],
+                 [89, 54, 156]])
+
+    def __init__(self,
+                 img_suffix='.jpg',
+                 seg_map_suffix='.png',
```
Comment (MengzhangLI) [inline mmseg/datasets/synapse.py:27]: In default setting, the label segmentation map would be `.jpg` rather than `.png`. [link of code](https://github.com/Dominic23331/mmsegmentation/blob/dev-1.x/tools/dataset_converters/synapse.py#L235-L236).

File: tools/dataset_converters/synapse.py (hunk 3 above)
Comment (MengzhangLI) [reply tools/dataset_converters/synapse.py:114]: The dataset split could be found here: https://github.com/Beckschen/TransUNet/tree/main/lists/lists_Synapse.

File: configs/_base_/datasets/synapse.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
+dataset_type = 'SynapseDataset'
```
Comment (MengzhangLI) [inline configs/_base_/datasets/synapse.py:1]: The setting of this synapse dataset config may be aligned with TransUNet default setting: https://github.com/Beckschen/TransUNet/blob/main/train.py#L19-L33.

**Comment by MengzhangLI** [comment]
First, we may use dataset from official repo, which can be downloaded by public http link, rather than TransUNet unofficial processed data.

Then we should ensure the our dataset conversion could handle dataset the same with [TransUNet](https://github.com/Beckschen/TransUNet/blob/main/datasets/README.md), `Convert them to numpy format, clip the images within [-125, 275], normalize each 3D image to [0, 1], and extract 2D slices from 3D volume for training cases while keeping the 3D volume in h5 format for testing cases.` The shape (H, W) and pixel value of each image should be the same. 

Next, we should check out normalization of TransUNet pretrained model, use its parameters in our backbone config, like [here](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/configs/_base_/models/upernet_swin.py#L4-L10).

Thanks for your nice PR again!

**Comment by MengzhangLI** [comment]
Could you please add doc for how to download the original data and how to convert it in [dataset_prepare.md](https://github.com/open-mmlab/mmsegmentation/blob/dev-1.x/docs/en/user_guides/2_dataset_prepare.md#prepare-datasets)?

File: tools/dataset_converters/synapse.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    mkdir_or_exist(osp.join(save_path, 'ann_dir/train'))
+    mkdir_or_exist(osp.join(save_path, 'ann_dir/val'))
+
+    for i, idx in enumerate(train_id):
+        img_3d = read_nii_file(
+            osp.join(dataset_path, 'img', 'img' + idx + '.nii.gz'))
+        label_3d = read_nii_file(
+            osp.join(dataset_path, 'label', 'label' + idx + '.nii.gz'))
+
+        img_3d = np.clip(img_3d, -125, 275)
```
Comment (MengzhangLI) [inline tools/dataset_converters/synapse.py:75]: Add comment above this line, give a link: https://github.com/Beckschen/TransUNet/tree/main/datasets to tell users why we clip (-125, 275). Thx.

**Comment by MengzhangLI** [comment]
This PR is wired to be closed automatically when I git push some configs for model training, @Dominic23331 could you please make a new PR agian? Sorry for my wrong operation!

Best,
